File: 4.8-bc/ts/stage-2-feature-first/src/features/billing/service.ts
// Billing feature — service: підписки.
import { randomUUID } from "node:crypto";

import type { Subscription } from "./model.js";
import type { Repository } from "./repository.js";

export class Service {
  constructor(private readonly repo: Repository) {}

  async subscribe(userId: string, plan: string): Promise<Subscription> {
    const now = new Date();
    const next = new Date(now);
    next.setMonth(next.getMonth() + 1);
    const sub: Subscription = {
      id: randomUUID(),
      userId,
      plan,
      nextChargeAt: next,
      createdAt: now,
    };
    await this.repo.create(sub);
    return sub;
  }
}

File: 4.8-bc/go/stage-2-feature-first/app/billing/service.go
// Package billing — Billing feature: підписки.
package billing

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type Service struct {
	repo *Repository
}

func NewService(repo *Repository) *Service {
	return &Service{repo: repo}
}

func (s *Service) Subscribe(ctx context.Context, userID uuid.UUID, plan string) (*Subscription, error) {
	now := time.Now().UTC()
	sub := Subscription{
		ID:           uuid.New(),
		UserID:       userID,
		Plan:         plan,
		NextChargeAt: now.AddDate(0, 1, 0),
		CreatedAt:    now,
	}
	if err := s.repo.Create(ctx, sub); err != nil {
		return nil, err
	}
	return &sub, nil
}
